File: knapsack/knapsack.rb
def file 
  @file ||= File.readlines(File.dirname(__FILE__) + "/knapsack1.txt").map { |x| x.gsub(/\n/, "") }
end

header = file.first
knapsack_size, number_of_items = header.split(" ")[0..1].map(&:to_i)

cache = [].tap { |m| (number_of_items+1).times { m << Array.new(knapsack_size+1) } }
cache[0].each_with_index { |value, weight| cache[0][weight] = 0  }

rows = file.drop(1).map { |row| row.split(" ").map(&:to_i)}

(1..number_of_items).each do |i|
  value, weight = rows[i-1]
  (0..knapsack_size).each do |x|
    # puts "weight: #{weight}, x: #{x}, i:#{i}"
    if weight > x
      cache[i][x] = cache[i-1][x] 
    else
      cache[i][x] = [cache[i-1][x], cache[i-1][x-weight] + value].max
    end
  end
end

# cache.each do |row|
#   p row
# end

p cache[number_of_items][knapsack_size]

File: knapsack/knapsack.py
import os

def knapsack_cached(rows, knapsack_size, index, cache):
    if(index is 0 or knapsack_size is 0):
        return 0
    else:
        value, weight = rows[index]
        if(weight > knapsack_size and knapsack_size not in cache[index-1]):
            cache[index-1][knapsack_size] = knapsack_cached(rows, knapsack_size, index-1, cache)
        else:
            if(knapsack_size not in cache[index-1]):
                option_1  = knapsack_cached(rows, knapsack_size, index-1, cache)
                option_2  = value + knapsack_cached(rows, knapsack_size - weight, index-1, cache)
                cache[index-1][knapsack_size] = max(option_1, option_2)                
            
        return cache[index-1][knapsack_size]
                    
file = open(os.path.dirname(os.path.realpath(__file__)) + "/knapsack2.txt")

knapsack_size, number_of_items = map(lambda x: int(x), file.readline().replace("\n", "").split(" "))

rows = []
for line in file.readlines():
    value, weight = line.split(" ")
    rows.append([int(value), int(weight)])

cache = []
for i in range(0, number_of_items):
    cache.append({})
    
result = knapsack_cached(rows, knapsack_size, number_of_items-1, cache)    
print(result)